proto/regen/ecocredit/v1alpha1/tx.proto

```diff
@@ -0,0 +1,167 @@
+
+// Msg is the regen.ecocredit.v1alpha1 Msg service.
+service Msg {
+
+    // CreateClass creates a new credit class with an approved list of issuers and optional metadata.
+    rpc CreateClass(MsgCreateClassRequest) returns (MsgCreateClassResponse);
+
+    // CreateBatch creates a new batch of credits for an existing credit class. This will create a new batch denom
+    // with a fixed supply. Issued credits can be distributed to recipients in either tradable or retired form.
+    rpc CreateBatch(MsgCreateBatchRequest) returns (MsgCreateBatchResponse);
+
+    // Send sends tradeable credits from one account to another account. Sent credits can either be tradable or retired on receipt.
+    rpc Send(MsgSendRequest) returns (MsgSendResponse);
+
+    // Retire retires a specified number of credits in the holder's account.
+    rpc Retire(MsgRetireRequest) returns (MsgRetireResponse);
+
+    // SetPrecision allows an issuer to increase the decimal precision of a credit batch. It is an experimental feature
+    // to concretely explore an idea proposed in https://github.com/cosmos/cosmos-sdk/issues/7113. The number of decimal
+    // places allowed for a credit batch is determined by the original number of decimal places used with calling CreatBatch.
+    // SetPrecision allows the number of allowed decimal places to be increased, effectively making the supply more
+    // granular without actually changing any balances. It allows asset issuers to be able to issue an asset without needing
+    // to think about how many subdivisions are needed upfront. While it may not be relevant for credits which likely have
+    // a fairly stable market value, I wanted to experiment a bit and this serves as a proof of concept for a broader
+    // bank redesign where say for instance a coin like the ATOM or XRN could be issued in its own units rather than
+    // micro or nano-units. Instead an operation like SetPrecision would allow trading in micro, nano or pico in the future
+    // based on market demand. Arbitrary, unbounded precision is not desirable because this can lead to spam attacks (like
+    // sending 0.000000000000000000000000000001 coins). This is effectively fixed precision so under the hood it is still
+    // basically an integer, but the fixed precision can be increased so its more adaptable long term than just an integer.
+    rpc SetPrecision(MsgSetPrecisionRequest) returns (MsgSetPrecisionResponse);
+}
+
+// MsgCreateClassRequest is the Msg/CreateClass request type.
+message MsgCreateClassRequest {
+
+    // designer is the address of the account which designed the credit class. The designer has special permissions
+    // to change the list of issuers and perform other administrative operations.
+    string designer = 1;
+
+    // issuers are the addresses of the approved issuers.
+    repeated string issuers = 2;
+
+    // metadata is any arbitrary metadata to attached to the credit class.
+    bytes metadata = 3;
```

Comment:
Do you think that there will be an `url` string in all classes metedata? If some, maybe it's worth to lift it up?

Reply:
likely, but here I'm trying to postpone dealing with data formats and just deal with asset mechanics. I'm assuming people will put some json here so this basically is the decision to start schema-less and iterate